src/Map.ts

```diff
@@ -0,0 +1,825 @@
+}
+
+/**
+ * @since 1.14.0
+ */
+const partitionMapWithKey = <K, RL, RR, A>(
+  fa: Map<K, A>,
+  f: (k: K, a: A) => Either<RL, RR>
+): Separated<Map<K, RL>, Map<K, RR>> => {
+  const left = new Map<K, RL>()
+  const right = new Map<K, RR>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    const ei = f(k, a)
+    if (ei.isLeft()) {
+      left.set(k, ei.value)
+    } else {
+      right.set(k, ei.value)
+    }
+  }
+  return {
+    left,
+    right
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+const partitionWithKey = <K, A>(fa: Map<K, A>, p: (k: K, a: A) => boolean): Separated<Map<K, A>, Map<K, A>> => {
+  const left = new Map<K, A>()
+  const right = new Map<K, A>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    if (p(k, a)) {
+      right.set(k, a)
+    } else {
+      left.set(k, a)
+    }
+  }
+  return {
+    left,
+    right
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+const filterMapWithKey = <K, A, B>(fa: Map<K, A>, f: (k: K, a: A) => Option<B>): Map<K, B> => {
+  const m = new Map<K, B>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    const o = f(k, a)
+    if (o.isSome()) {
+      m.set(k, o.value)
+    }
+  }
+  return m
+}
+
+/**
+ * @since 1.14.0
+ */
+const filterWithKey = <K, A>(fa: Map<K, A>, p: (k: K, a: A) => boolean): Map<K, A> => {
+  const m = new Map<K, A>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [k, a] = e.value
+    if (p(k, a)) {
+      m.set(k, a)
+    }
+  }
+  return m
+}
+
+/**
+ * Create a Map from a foldable collection of key/value pairs, using the
+ * specified function to combine values for duplicate keys.
+ *
+ * @since 1.14.0
+ */
+export function fromFoldable<K, F extends URIS3>(
+  S: Setoid<K>,
+  F: Foldable3<F>
+): <U, L, A>(ta: Type3<F, U, L, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F extends URIS2>(
+  S: Setoid<K>,
+  F: Foldable2<F>
+): <L, A>(ta: Type2<F, L, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F extends URIS>(
+  S: Setoid<K>,
+  F: Foldable1<F>
+): <A>(ta: Type<F, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F>(
+  S: Setoid<K>,
+  // tslint:disable-next-line: deprecation
+  F: Foldable<F>
+): <A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A>
+export function fromFoldable<K, F>(
+  S: Setoid<K>,
+  // tslint:disable-next-line: deprecation
+  F: Foldable<F>
+): <A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => Map<K, A> {
+  return <A>(ta: HKT<F, [K, A]>, f: (existing: A, a: A) => A) => {
+    return F.reduce<[K, A], Map<K, A>>(ta, new Map<K, A>(), (b, [k, a]) => {
+      const bOpt = lookupWithKey(S)(k, b)
```

Comment:
`lookupWithKey(S)` can be moved outside

Reply:
I made this change. Marking it as resolved.